README
Copyright 2008, 2009 Kevin Ryde

This file is part of Perl-Critic-Pulp.

Perl-Critic-Pulp is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 3, or (at
your option) any later version.

Perl-Critic-Pulp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with Perl-Critic-Pulp.  If not, see <http://www.gnu.org/licenses/>.



This is some add-on policies for Perl::Critic, for miscellaneous bugs,
compibility, efficiency and some cosmetic things.  See the POD of
Perl::Critic::Pulp for the full list.

The only major prerequisite is Perl::Critic itself, though
"PerlMinimumVersionAndWhy" does nothing without Perl::MinimumVersion.

The Perl-Critic-Pulp home page is

    http://user42.tuxfamily.org/perl-critic-pulp/index.html


Changes
Copyright 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2019, 2021 Kevin Ryde

This file is part of Perl-Critic-Pulp.

Perl-Critic-Pulp is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 3, or (at
your option) any later version.

Perl-Critic-Pulp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with Perl-Critic-Pulp.  If not, see <http://www.gnu.org/licenses/>.



Version 98, February 2021
    - ProhibitUnknownBackslash new charnames option

Version 97, October 2019
    - RequireFinalSemicolon allow hashref in arrayref (which PPI 1.270
      says code block), as reported by Andy Lester RT#130725

Version 96, December 2017
    - ProhibitUnknownBackslash allow for non-ascii in strings, as
      reported by Mark Fowler RT#123922

Version 95, October 2017
    - RequireFinalSemicolon also sort{} blocks are an expression

Version 94, June 2017
    - ProhibitUnknownBackslash \N also Perl 5.16, as reported by
      Ed Davis RT#122064

Version 93, April 2017
    - some doubtfulness in Makefile.PL meta bits

Version 92, November 2016
    - ProhibitPOSIXimport oops, don't use // operator, per paul RT#118959

Version 91, November 2016
    - RequireFinalCut allow no blank line before =cut, as reported by
      Andy Lester RT#118722
    - RequireLinkedURLs allow example.com

Version 90, March 2015
    - PerlMinimumVersionAndWhy actually check Modern::Perl
    - ProhibitNullStatements beware of extra stuff after try/catch

Version 89, January 2015
    - PerlMinimumVersionAndWhy experiment Modern::Perl meaning 5.10,
      as proposed by Ed Avis RT#100290
    - RequireFinalSemicolon recognise try/catch of Try.pm, TryCatch.pm
      and Syntax::Feature::Try, as proposed by Ed Avis RT#100291.
      Various List::Util functions as expression blocks.
    - RequireTrailingCommaAtNewline exception for single here-documents,
      as proposed by Ed Avis RT#100292
    - ProhibitNullStatements report null ; on try/catch, and look
      inside try/catch/finally blocks

Version 88, November 2014
    - ConstantBeforeLt oops again, PPI incompatibility too much, must
      do and depend on its 1.220

Version 87, November 2014
    - ConstantBeforeLt really allow for PPI

Version 86, November 2014
    - ConstantBeforeLt allow for PPI incompatible change to ->prototype()

Version 85, May 2014
    - oops forgot Pod::Escapes in PREREQ_PM, per Petr Pisar RT#95440

Version 84, May 2014
    - new Documentation::RequireFilenameMarkup

Version 83, May 2014
    - PerlMinimumVersionAndWhy new 5.10 -e -x stacked filetest per
      Kent Fredric RT#95208

Version 82, April 2014
    - ProhibitAdjacentLinks, ProhibitLinkToSelf,
      ProhibitParagraphTwoDots, ProhibitUnbalancedParens,
      ProhibitVerbatimMarkup, RequireLinkedURLs check =begin blocks
      only for POD-type ":"
    - ProhibitBadAproposMarkup check =begin :man and :roff only
    - ProhibitDuplicateSeeAlso ignore all =begin blocks

Version 81, April 2014
    - ProhibitAdjacentLinks fix warnings when internal and external
      links adjacent, as reported by Mike O'Regan RT#94318
    - ConstantPragmaHash and ConstantLeadingUnderscore pass ->content
      to version.pm as it enquires too much into a PPI object ref,
      per RT#92100 dmacks and ppisar

Version 80, July 2013
    - new policy Documentation::ProhibitParagraphEndComma
    - tests UnexpandedSpecialLiteral.t don't exercise __PACKAGE__ =>
      across newline since that might be going to change in Perl 5.20,
      as reported by Reini Urban RT#87307

Version 79, March 2013
    - fix ProhibitFatCommaNewline load version.pm for perl 5.8, as
      reported by paul RT#83990

Version 78, March 2013
    - new policies CodeLayout::ProhibitIfIfSameLine,
                   CodeLayout::ProhibitFatCommaNewline

Version 77, February 2013
    - new policy Documentation::ProhibitDuplicateHeadings
    - fix POD parse losing violations across =cut gaps

Version 76, January 2013
    - RequireNumericVersion explicit check for exponential strings "1e6" 

Version 75, November 2012
    - ProhibitEmptyCommas fix for [{%a},{}] reported by Mike O'Regan
      RT#81390

Version 74, October 2012
    - tests RequireNumericVersion skip version.pm 1e6 uncertain quite yet

Version 73, June 2012
    - tests fix for File::Temp in perl pre-5.10

Version 72, June 2012
    - TextDomainPlaceholders fix #line regexp plain \2 for perl before 5.10

Version 71, June 2012
    - RequireLinkedURLs - skip "=begin html" etc blocks
    - RequireFinalCut - cut not required if only =begin foo/=end foo block

Version 70, May 2012
    - new policy Documentation::RequireFinalCut

Version 69, January 2012
    - new policy ValuesAndExpressions::ProhibitDuplicateHashKeys
    - PerlMinimumVersionAndWhy
      - new keys @array, values @array and each @array are 5.12
      - new eval '#line' with #line the very first thing is 5.8
    - Perl::Critic::Pulp::Utils new elem_is_comma_operator()

Version 68, January 2012
    - RequireNumericVersion
      - recognise fully-qualified $Foo::VERSION
      - recognise "$VERSION = eval $VERSION" trick

Version 67, December 2011
    - RequireTrailingCommaAtNewline allow "return(123 newline );"

Version 66, December 2011
    - new policy Documentation::ProhibitDuplicateSeeAlso

Version 65, September 2011
    - PerlMinimumVersionAndWhy
      - new UNIVERSAL.pm methods in 5.004 and 5.10
      - new "Foo->$method" no-args without parens is 5.6 

Version 64, August 2011
    - fix ProhibitLinkToSelf warnings on no L<> display part

Version 63, August 2011
    - new policy Documentation::ProhibitLinkToSelf

Version 62, July 2011

Makefile.PL
#!/usr/bin/perl -w

# Copyright 2008, 2009, 2010, 2011, 2012, 2014, 2015 Kevin Ryde

# This file is part of Perl-Critic-Pulp.

# Perl-Critic-Pulp is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 3, or (at your option) any
# later version.
#
# Perl-Critic-Pulp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with Perl-Critic-Pulp.  If not, see <http://www.gnu.org/licenses/>.

use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile
  (NAME         => 'Perl::Critic::Pulp',
   ABSTRACT     => 'Some add-on policies for Perl::Critic.',
   VERSION_FROM => 'lib/Perl/Critic/Pulp.pm',
   AUTHOR       => 'Kevin Ryde <user42_kevin@yahoo.com.au>',
   LICENSE      => 'gpl_3',
   SIGN         => 1,
   MIN_PERL_VERSION => '5.006',

   PREREQ_PM => {
                 # 1.220 for its incompatible PPI::Statement::Sub->prototype()
                 'PPI' => '1.220',
                 'PPI::Document' => 0,
                 # actually PPI::Dumper only used in the .t scripts if
                 # there's an error ...
                 'PPI::Dumper' => 0,

                 # 1.082 or some such for new style parameter specs
                 # 1.084 for highest_explicit_perl_version() method
                 'Perl::Critic' => '1.084',
                 # parts of Perl::Critic, used so name them explicitly
                 'Perl::Critic::Policy' => '1.084',
                 # Utils 1.100 for precedence_of() supporting -f etc filetests
                 'Perl::Critic::Utils' => '1.100',
                 'Perl::Critic::Utils::PPI' => 0,
                 'Perl::Critic::Violation' => 0,

                 # Required by Perl::Critic itself anyway.
                 # 1.02 for avoiding leaks in perl 5.8, probably
                 # something newer has more merit though
                 'IO::String' => '1.02',

                 # Also used by Perl::Critic itself.  Maybe should
                 # require 0.24 for various XS bugfixes.
                 'List::MoreUtils' => '0.24', # 0.24 for stack growing, maybe

                 # Could make this optional, but it doesn't drag in anything
                 # outrageous.
                 'Pod::MinimumVersion' => 50, # v.50 fix text in =item etc

                 # Builtin in recent perl anyway.
                 'List::Util' => 0,
                 'Test::More' => 0,
                 'Scalar::Util' => 0,

                 # Required by Perl::Critic itself anyway,
                 # and comes with perl 5.10 anyway too.
                 'version' => 0,

                 'Pod::Escapes' => 0,
                 'Pod::Parser' => 0,
                 'Pod::ParseLink' => 0,
                },

   META_MERGE =>
   { resources =>
     { homepage => 'http://user42.tuxfamily.org/perl-critic-pulp/index.html',
       license  => 'http://www.gnu.org/licenses/gpl.html',
     },
     # don't index "package Perl::MinimumVersion" additions there
     no_index => { package => [ 'Perl::MinimumVersion' ],
                   directory=>['devel','xt'],
                 },

     recommends => { 'Perl::MinimumVersion' => 0,
                   },
     prereqs =>
     { test =>
       { suggests =>
         { 'Perl::MinimumVersion' => 0,
         },
       },
     },
     optional_features =>
     { maximum_policies =>
       { description => 'Have the policies check as much as possible.',
         requires => { 'Perl::MinimumVersion' => 0,
                     },
       },
     },
   },
  );


# maximum_devel =>
# { description => 'Stuff used variously for development.',
#   requires => { 'PPI::Document' => 0,
#                 'Perl::MinimumVersion' => 0,
#                 'Encode' => 0,
#                 'Perl6::Slurp' => 0,
#                 'Regexp::Common' => 0,
#                 'Devel::Mallinfo' => 0,
#                 'IO::Uncompress::AnyInflate' => 0,
#                 'lib::abs' => 0,
#                 'IO::String' => 0,
#                 'Digest::MD5' => 0,
#                 'I18N::Langinfo' => 0,
#                 'Locale::TextDomain' => 0,
#                 'File::Locate' => 0,
#
#                 'File::Locate::Iterator' => 0, # mine
#                 # parts of File::Locate::Iterator
#                 'Iterator::Simple' => 0,
#                 'Iterator::Simple::Locate' => 0,
#               },
# },
